// @runanywhere/web-llamacpp/dist/Extensions/DiffusionTypes.js
/** RunAnywhere Web SDK - Diffusion Types */
export var DiffusionScheduler;
(function (DiffusionScheduler) {
    DiffusionScheduler[DiffusionScheduler["DPM_PP_2M_Karras"] = 0] = "DPM_PP_2M_Karras";
    DiffusionScheduler[DiffusionScheduler["DPM_PP_2M"] = 1] = "DPM_PP_2M";
    DiffusionScheduler[DiffusionScheduler["DPM_PP_2M_SDE"] = 2] = "DPM_PP_2M_SDE";
    DiffusionScheduler[DiffusionScheduler["DDIM"] = 3] = "DDIM";
    DiffusionScheduler[DiffusionScheduler["Euler"] = 4] = "Euler";
    DiffusionScheduler[DiffusionScheduler["EulerAncestral"] = 5] = "EulerAncestral";
    DiffusionScheduler[DiffusionScheduler["PNDM"] = 6] = "PNDM";
    DiffusionScheduler[DiffusionScheduler["LMS"] = 7] = "LMS";
})(DiffusionScheduler || (DiffusionScheduler = {}));
export var DiffusionModelVariant;
(function (DiffusionModelVariant) {
    DiffusionModelVariant[DiffusionModelVariant["SD_1_5"] = 0] = "SD_1_5";
    DiffusionModelVariant[DiffusionModelVariant["SD_2_1"] = 1] = "SD_2_1";
    DiffusionModelVariant[DiffusionModelVariant["SDXL"] = 2] = "SDXL";
    DiffusionModelVariant[DiffusionModelVariant["SDXL_Turbo"] = 3] = "SDXL_Turbo";
    DiffusionModelVariant[DiffusionModelVariant["SDXS"] = 4] = "SDXS";
    DiffusionModelVariant[DiffusionModelVariant["LCM"] = 5] = "LCM";
})(DiffusionModelVariant || (DiffusionModelVariant = {}));
export var DiffusionMode;
(function (DiffusionMode) {
    DiffusionMode[DiffusionMode["TextToImage"] = 0] = "TextToImage";
    DiffusionMode[DiffusionMode["ImageToImage"] = 1] = "ImageToImage";
    DiffusionMode[DiffusionMode["Inpainting"] = 2] = "Inpainting";
})(DiffusionMode || (DiffusionMode = {}));
